### PR TITLE
Allow setting PreferredDuringSchedulingIgnoredDuringExecution anti-af…

### PR DESCRIPTION
**Pull request by EvanKrall**
…finity terms with anti_affinity_preferred key.

### CODE REVIEW

File: paasta_tools/kubernetes_tools.py [hunk 1]
```diff
@@ -1937,9 +1934,28 @@ def get_pod_anti_affinity(self) -> Optional[V1PodAntiAffinity]:
...
+        affinities with the "paasta.yelp.com" prefixed label selector
+        :return:
+        """
+
+        required_terms = self.get_pod_anti_affinity_terms(
+            config_key="anti_affinity",
+        )
+        preferred_terms = self.get_pod_anti_affinity_terms(
+            config_key="anti_affinity_preferred",
+        )
```
Comment (nemacysts) [inline paasta_tools/kubernetes_tools.py:1951]: i think ideally we'd have something that looked like:
```YAML
anti_affinity:
    preferred:
        - ...
    required:
        - ...
```

but we probably don't wanna deal with doing the migration to a new config format right now :p

File: paasta_tools/kubernetes_tools.py [hunk 2]
```diff
@@ -1911,13 +1911,10 @@ def get_node_affinity(self) -> Optional[V1NodeAffinity]:
...
 
-    def get_pod_anti_affinity(self) -> Optional[V1PodAntiAffinity]:
-        """
-        Converts the given anti-affinity on service and instance to pod
-        affinities with the "paasta.yelp.com" prefixed label selector
-        :return:
-        """
-        conditions = self.config_dict.get("anti_affinity", [])
+    def get_pod_anti_affinity_terms(
+        self, config_key: str
```
Comment (nemacysts) [inline paasta_tools/kubernetes_tools.py:1915]: we can probably type config_key as a Literal, but that's probably not worth it atm

**Review by nemacysts — APPROVED** [review]

File: paasta_tools/kubernetes_tools.py (hunk 2 above)
Comment (EvanKrall) [reply paasta_tools/kubernetes_tools.py:1915]: easy enough, and it caught the fact that I never defined a type in `KubernetesDeploymentConfigDict` for `anti_affinity_preferred`.

File: paasta_tools/kubernetes_tools.py [hunk 3]
```diff
@@ -1937,9 +1934,28 @@ def get_pod_anti_affinity(self) -> Optional[V1PodAntiAffinity]:
...
+            config_key="anti_affinity_preferred",
+        )
+
+        if required_terms is None and preferred_terms is None:
+            return None
 
         return V1PodAntiAffinity(
-            required_during_scheduling_ignored_during_execution=affinity_terms
+            required_during_scheduling_ignored_during_execution=required_terms,
+            preferred_during_scheduling_ignored_during_execution=preferred_terms,
```
Comment (EvanKrall) [inline paasta_tools/kubernetes_tools.py:1958]: d'oh, preferred_during_scheduling_ignored_during_execution needs to be a list of V1WeightedPodAffinityTerm, not V1PodAffinityTerm. I'm gonna have to un-DRY this.

Comment (nemacysts) [reply paasta_tools/kubernetes_tools.py:1958]: heh, i guess that my assumption that there'd be a default weight and we could use the same class for both anti-affinities was wrong :p

**Review by nemacysts — APPROVED** [review]

**Comment by EvanKrall** [comment]
@nemacysts how do you feel about this? Is this still worth shipping? COMPINFRA-1211

**Comment by nemacysts** [comment]
> @nemacysts how do you feel about this? Is this still worth shipping? COMPINFRA-1211

@EvanKrall I think this has come up enough times that it's probably still worth having available - I do feel a little iffy about having people add this willy-nilly to their configs, but we can either use dyncow to require our approval OR just not publicize/document this too much :)

**Review by nemacysts — APPROVED** [review]

**Comment by gmdfalk** [comment]
> > @nemacysts how do you feel about this? Is this still worth shipping? COMPINFRA-1211
> 
> @EvanKrall I think this has come up enough times that it's probably still worth having available - I do feel a little iffy about having people add this willy-nilly to their configs, but we can either use dyncow to require our approval OR just not publicize/document this too much :)

Agree, i think not advertising it will be okay and we could test the feature itself on paasta-contract-monitor.

When i release https://github.com/Yelp/paasta/pull/3512 next week, should i roll this one in as well?

**Pull request merged**